### ISSUE TITLE
⚡️ 0.28 RELEASE 🚀

### ISSUE THREAD

**Issue by vasco-santos** [libp2p/js-libp2p#603]
> The peer-store revamp with an easier management of peers' data and persistence!

# 🗺 What's left for release

**`js-ipfs`** integration
- [x] [ipfs/js-ipfs-bitswap#217](https://github.com/ipfs/js-ipfs-bitswap/pull/217)
- [ ] [ipfs/js-ipfs#3019](https://github.com/ipfs/js-ipfs/pull/3019)

# 🔦 Highlights

## 📜 Improved PeerStore towards v2

Libp2p's PeerStore is responsible for keeping an updated register with the relevant information of the known peers. It should be the single source of truth for all peer data, where a subsystem can learn about peers' data and where someone can listen for data updates. 

The initial design of the PeerStore (formerly known as [`peer-book`](https://github.com/libp2p/js-peer-book)) was not implemented with some functionalities in mind that we plan to land over the next releases, so that node connectivity and connection management are improved.

The PeerStore will now comprise four main components: `addressBook`, `keyBook`, `protoBook` and `metadataBook`. With this separation of concerns, libp2p now provides a PeerStore API so that known data of a peer can be easily added and a persistence layer, in order to enable faster connectivity on subsequent node starts, since it will not to discover the stored peers one more time to establish a connection with them.

With this revamp of the PeerStore, the building blocks for the PeerStore v2 are created. As a result, we will be able to focus on the PeerStore v2 over the coming releases, which might add multiaddr confidence, peer scoring, among other improvements aiming to improve connectivity efficiency and connection management.

## ♻️ Removed peer-info abstraction layer

[`PeerInfo`](https://github.com/libp2p/js-peer-info) is a libp2p peer abstraction layer that combines a [`PeerId`](https://github.com/libp2p/js-peer-id) with known data of the peer, namely its multiaddrs and protocols. It has been used for a long time by `js-libp2p` and its modules to carry this data around the libp2p stack, as well as by the libp2p API, both for providing this data to the users or to receive it from them.

Since this PeerInfo instances were navigating through the entire codebases, some data inconsistencies could be observed in libp2p. Different libp2p subsystems were running with different visions of the known peers data. For instance, a libp2p subsystem receives a copy of this instance with the peer multiaddrs and protocols, but if new data of the peer is obtained from other subsystem, it would not be updated on the former. Moreover, considering that several subsystems were modifying the peer data, libp2p had no visibility on which data was the updated and the older one.

Considering the complete revamp of the libp2p PeerStore towards its second version. With this new PeerStore acting as the single source of truth, we do not need to carry [`PeerInfo`][peer-info] instances around. This also solves all the problems stated above, since subsystems will report new observations to the PeerStore and it will provide clear and updated picture of the known peers' data.

Given that [`peer-info`](https://github.com/libp2p/js-peer-info) is being deprecated, all the API methods with parameters or return values were changed in favour of the `peer-id`. If someone need more information about the peer, it is now recommended to use the PeerStore API.

## ⚙️ Address Manager

With the deprecation of the `peer-info` module, `js-libp2p` must provide a new way of specifying which multiaddrs should be used by its underlying transports to listen for new connections. With this in mind and some other discussions regarding supporting `announce` and `noAnnounce` addresses in the past, we created the Address Manager.

With this release, `js-libp2p` supports the configuration of `listen`, `announce` and `noAnnounce` multiaddrs. These will be used to create the Address Manager (`listen`) and by other subsystems which announce the peer addresses to the outside world (`announce` and `noAnnounce`).

## 🛡️Full support for ed25519 keys and noise handshake protocol

`js-ibp2p` now fully supports `ed25519` keys and JS nodes can establish connections with GO nodes using them!

Thanks to NodeFactoryIo, `js-libp2p` now supports the noise handshake protocol through [NodeFactoryIo/js-libp2p-noise](https://github.com/NodeFactoryIo/js-libp2p-noise) module. This module is now fully supported and tested through the libp2p codebase and is interoperable with the noise implementation for `go-libp2p`.

## 🔒Out of the box keychain

Libp2p already provided the [libp2p-keychain](https://github.com/libp2p/js-libp2p-keychain) module that users could use for key management and cryptographically protected messages. However, users would need to properly setup the keychain as a separate piece of the stack.

In `0.28`, libp2p keychain is now part of the libp2p core 🎉.  As a result, libp2p users can now provide the keychain configuration through the libp2p config and use the keychain via its API, i.e `libp2p.keychain`.

# 🏗 API Changes

See the [API.md][api] readme for detailed usage on the new API and the [migration guide v0.27 to v0.28](https://github.com/libp2p/js-libp2p/blob/docs/migration-0.27-to-0.28/doc/migrations/v0.27-v.28.md). 

Significant breaking changes are briefly listed below:

- All peerInfo usage
- libp2p.peerStore

# ✅ Release Checklist

- Robustness and quality
  - [x] Ensure that all tests are passing, this includes:
    - [x] unit
  - [x] Publish a release candidate to npm
    ```sh
    # Minor prerelease (e.g. 0.24.1 -> 0.25.0-rc.0)
    $ npx aegir release --type preminor -t node -t browser --preid rc --dist-tag next

    # Increment prerelease (e.g. 0.25.0-rc.0 -> 0.25.0-rc.1)
    $ npx aegir release --type prerelease -t node -t browser --preid rc --dist-tag next
    ```
  - [ ] Run tests of the following projects with the new release:
    - [ ] [js-ipfs](https://github.com/ipfs/js-ipfs)
- Documentation
  - [x] Ensure that README.md is up to date
  - [x] Ensure that all the examples run
- Communication
  - [ ] Create the release issue
  - [ ] Take a snapshot between of everyone that has contributed to this release (including its subdeps in IPFS, libp2p, IPLD and multiformats) using [`name-your-contributors`](https://www.npmjs.com/package/name-your-contributors). Generate a nice markdown list with [this script](https://gist.github.com/jacobheun/d2ff479ca991733c13cdcf688a1317e5)
  - [ ] Announcements (both pre-release and post-release)
    - [ ] Twitter
    - [ ] IRC
    - [ ] Reddit
    - [ ] [discuss.libp2p.io](https://discuss.libp2p.io/c/news)
  - [ ] Blog post
  - [ ] Copy release notes to the [GitHub Release description](https://github.com/libp2p/js-libp2p/releases)

# ❤️ Huge thank you to everyone that made this release possible

TODO

# 🙌🏽 Want to contribute?

Would you like to contribute to the libp2p project and don't know how? Well, there are a few places you can get started:

- Check the issues with the `help wanted` label in the [libp2p repo](https://github.com/libp2p/js-libp2p/issues?q=is%3Aopen+is%3Aissue+label%3A%22help+wanted%22)
- Join an IPFS All Hands, introduce yourself and let us know where you would like to contribute - https://github.com/ipfs/team-mgmt#all-hands-call
- Hack with IPFS and show us what you made! The All Hands call is also the perfect venue for demos, join in and show us what you built
- Join the discussion at http://discuss.libp2p.io/ and help users finding their answers.
- Join the [⚡️libp2p Weekly Sync 🙌🏽](https://github.com/libp2p/team-mgmt/issues/16) and be part of the Sprint action!

# ⁉️ Do you have questions?

The best place to ask your questions about libp2p, how it works and what you can do with it is at [discuss.libp2p.io](https://discuss.libp2p.io). We are also available at the #libp2p channel on Freenode.

[api]: https://github.com/libp2p/js-libp2p/blob/master/doc/API.md
[config]: https://github.com/libp2p/js-libp2p/blob/master/doc/CONFIGURATION.md
[docs]: https://github.com/libp2p/js-libp2p/blob/master/doc/
[discuss]: https://discuss.libp2p.io

**Comment by jacobheun** [comment]
FYI: All code slated for the 0.28 release will be merged into the [0.28.x](https://github.com/libp2p/js-libp2p/tree/0.28.x) branch during development.

**Comment by vasco-santos** [comment]
After release checklist:

- [ ] change libp2p modules npm tags from `beta` to `latest`
- [ ] deprecate `peer-info`
- [ ] deprecate `libp2p-kecyhain`

**Comment by jacobheun** [comment]
The initial rc **0.28.0-rc.0** is out, it can be installed via `npm install libp2p@next`. 

Please keep in mind that dependencies that required breaking changes due to the removal of `PeerInfo` will need to be installed with via their beta tag, such as `npm install libp2p-webrtc-star@beta`. You can check specific modules via `npm show` (`npm show libp2p-webrtc-star`) to see if a beta version exists, if so you should use that for 0.28. Once the release is ready we will change the beta version to latest for all affected modules.

**Comment by jacobheun** [comment]
[0.28.0](https://github.com/libp2p/js-libp2p/releases/tag/v0.28.0) has been released. We've also updated npm dist-tags for everything that had a breaking change due to PeerInfo being removed, so `npm install` will now install modules compliant with 0.28.
